src/imitation/data/types.py

```diff
@@ -22,7 +29,267 @@
+    ) -> "DictObs":
+        """Indexes or slices each array.
+
+        See `.get` for accessing a value from the underlying dictionary.
+
+        Note that it will still return singleton values as np.arrays, not scalars,
+        to be consistent with DictObs type signature.
+
+        Args:
+            key: a single slice
+
+        Returns:
+            A new DictObj object with each array indexed.
+        """
+        # asarray handles case where we slice to a single array element.
+        return self.__class__({k: np.asarray(v[key]) for k, v in self._d.items()})
+
+    def __iter__(self) -> Iterator["DictObs"]:
+        """Iterates over the first dimension of each array.
+
+        Raises:
+            ValueError if len() is not defined.
+
+        Returns:
+            Iterator of dictobjs by first dimension.
+        """
+        return (self[i] for i in range(len(self)))
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        if not self.keys() == other.keys():
+            return False
+        return all(np.array_equal(self.get(k), other.get(k)) for k in self.keys())
+
+    @property
+    def shape(self) -> Dict[str, Tuple[int, ...]]:
+        """Returns a dictionary with shape-tuples in place of the arrays."""
+        return {k: v.shape for k, v in self.items()}
+
+    @property
+    def dtype(self) -> Dict[str, np.dtype]:
+        """Returns a dictionary with dtype-tuples in place of the arrays."""
+        return {k: v.dtype for k, v in self.items()}
+
+    def keys(self):
+        return self._d.keys()
+
+    def values(self):
+        return self._d.values()
+
+    def items(self):
+        return self._d.items()
+
+    def __contains__(self, key):
+        return key in self._d
+
+    def get(self, key: str) -> np.ndarray:
+        """Returns the array for the given key, or raises KeyError."""
+        return self._d[key]
+
+    def unwrap(self) -> Dict[str, np.ndarray]:
+        """Returns a copy of the underlying dictionary (arrays are not copied)."""
+        return {k: v for k, v in self._d.items()}
+
+    def map_arrays(self, fn: Callable[[np.ndarray], np.ndarray]) -> "DictObs":
+        """Returns a new DictObs with `fn` applied to every array."""
+        return self.__class__({k: fn(v) for k, v in self.items()})
+
+    @staticmethod
+    def _unravel(dictobs_list: Iterable["DictObs"]) -> Dict[str, List[np.ndarray]]:
+        """Converts a list of DictObs into a dictionary of lists of arrays."""
+        it1, it2 = itertools.tee(dictobs_list)
+        # assert all have same keys
+        key_set = set(frozenset(obs.keys()) for obs in it1)
+        if len(key_set) == 0:
+            raise ValueError("Empty list of DictObs")
+        if not len(key_set) == 1:
+            raise ValueError(f"Inconsistent keys: {key_set}")
+
+        unraveled: Dict[str, List[np.ndarray]] = collections.defaultdict(list)
+        for do in it2:
```

Comment:
Nit: `do` not obviously an abbreviation for dict obs rather than to "do" something.